### ISSUE TITLE
Prune feature?

### ISSUE THREAD

**Issue by brainchild0** [sirixdb/sirix#562]
I understand SirixDB is designed for append only. Old versions of a document remain, but are shadowed in queries unless they express a temporal dimension.

The approach is valuable in some ways, but  capacity limitations must be considered. For example, on an embedded environment, cleaning old revisions may be needed for freeing space.  For client-side applications, it may be infeasible to maintain  a database that  grows consistently and indefinitely. 

I tend to understand the design of the storage layer as being optimized for append, both for write latency and storage optimization. It may be a feature, or it may be an anti-feature. For an append-only system, rebuilding a new database is the only way to prune. Even a bad internal pruning feature is unlikely to be worse than the alternative of bypassing the database manager features, and writing document copies to a fresh database instance. Therefore, even a solution might be useful of simply rebuilding a new internal representation from scratch, and dropping the old representation, within the same database file.

Assuming I understand correctly that no pruning feature is available, is it possible to consider for the future?

From an application standpoint, it may be desired to remove revisions older than a certain date (e.g. 1 year), or older than a certain revision count (e.g. 100 revisions). It may also be desired to prune according to frequency (e.g. erase all documents older than 3 years, keep one each month for range of 1 to 3 years old).

**Comment by JohannesLichtenberger** [comment]
Currently, we store the JSON or XML resources in a binary representation in one resource file per JSON file plus a second offset file into the revisions. On the one hand, we'd have to make sure that there is no read-only trx open on old revisions (which should be easy to track). On the other hand, we'd have to track which revision is the oldest we can prune (due to sliding snapshot versioning of the data pages itself -- only writing changed nodes plus nodes, which fall out of a sliding window; thus if a database page fragment has not been changed in the last couple of revisions, but then nodes in the page change, we could only prune up to this old revision - 1). To not mess up all offset references, we'd have to punch "a whole" into the file after the headers/UberPages are written. I guess most filesystems should be able to not physically store these 0-bytes (sparse files).

Due to the clever sliding snapshot algorithm, the pruning is not trivial, at least, as the inner page pointing to the leaf data page(s) might have several references pointing to more than one page fragments (if the page has been changed) up to a predefined threshold.

A no-brainer solution for this feature would be to store the data file (it's an indexed log) in another log-structured storage system (for instance, Kafka or Pulsar/Bookkeeper?). However, I'd like to make these additional, mainly distributed backends optional. We could also, for instance, store on a local disk plus in a Kafka cluster or something like that.

**Comment by brainchild0** [comment]
> To not mess up all offset references, we'd have to punch "a whole" into the file after the headers/UberPages are written. I guess most filesystems should be able to not physically store these 0-bytes (sparse files).

Yes, but I think what is really desired is a truly compacted file. If the storage layer cannot support offset rewriting, then perhaps it is worth considering a possibility closer to my suggestion, of rebuilding a new copy from scratch, just doing it such that the client is insulated through the API from the heavy lifting.

Although it would not be backward compatible, one way to handle compaction would be through an intermediary table that translates offsets from a physical to virtual space, much like the way a kernel handles memory paging.

**Comment by brainchild0** [comment]
I'm not familiar with the fine details of the system, but please let me know whether the following is of any help.

For document revisions, *A*, *B*, and *C*, a simple delta-only model of storing revisions results in three delta operations, *delta(null, A)*, *delta(A, B)*, and *delta(B, C)*. It may be difficult to prune *B*, because it is needed to generated *C*. However, if we considered a less efficient storage model, of storing *delta(null, A)*, *delta(A, B)*, and *delta(A, C)*, then *B* may be pruned without dependencies. If we generalize to a full binary tree representing an arbitrary revision count, built by  breadth-first insertion, then we may be able to erase parts of the history simply by clearing complete subtrees. It is not perfect. Some revisions will need to be flagged as "deleted", though physically preserved.

Storage efficiency is hybrid between the most compact model targeting append only, and a basic model of writing a full document to unallocated space.

I understand I may be using the wrong concepts to describe your  storage model, because I am thinking at a very generic level, not knowing most of the relevant details.

**Comment by JohannesLichtenberger** [comment]
Generally, it's a fully persistent data structure (tree of tries for 64bit int/long key mapping to node values). That is, the primary document index trie + tries to store the secondary, user-defined indexes.
In one file, offsets into the revision root pages are stored (plus timestamps to do an in-memory binary search if a revision is opened by its timestamp). In the second file, the data itself is appended. SirixDB does a copy-on-write of database pages and has to copy the paths up to the `RevisionRootPage` if a data page fragment changes (the leaf nodes in the tree). There's also the little twist that SirixDB does not simply copy the data page itself, but only nodes that have changed plus nodes that fall out of a sliding window (inner index pages only have references with offsets into the file plus checksums plus maybe in the future an encryption key and are thus relatively lightweight (additionally drawing some inspiration from adaptive radix tries and array mapped tries to reduce the sizes further of the inner nodes -- storing only up to 4 keys and a bitmap representation up to the full references to 1024 child pages).

I've added some images, which may hint at how data is stored in SirixDB, but I wanted to publish another article with more details (the devices in the on-disk layout are instead 2 files).

I think what's currently missing is that nodes can be inserted with a given `nodeKey` instead of provided by a simple `AtomicLong` sequence generator.

- First, a new revision has to be written into a temp directory based on a given revision number or timestamp (timestamp might also be now - some time). I think for the JSON layer, maybe copy methods to read from a read-only trx are currently missing
- Second, SirixDB has to read the revision diffs (currently, IIRC, they are stored for the JSON resources and not XML, but that would be a significant -- maybe rather easy - contribution, as it probably can be abstracted ;-)).
- For every revision up to the most recent, these revision diffs have to be read, and data has to be inserted with the nodeKeys stored in the diffs.
- After every revision a commit has to be issued. In that way, we'd preserve the whole revision history from the point in time where the old revisions have to be pruned up to the most recent (and we do not have to cope with the storage internals)
- Afterwards the old directory has to be deleted and the current temp directory has to be renamed.
- We also have to make sure to copy index definitions (the `indexes` folder to the new resource).

![json-tree-encoding](https://user-images.githubusercontent.com/1202851/205322956-8fb44130-9488-4998-a9cb-11fcf80639ab.png)
![logical-on-device-layout](https://user-images.githubusercontent.com/1202851/205322962-4f54c117-b197-4e94-aecf-0770410b2f55.png)
![sirix-resource-revisions](https://user-images.githubusercontent.com/1202851/205322968-a9888a4c-34af-4041-a97f-a23cdb6115c7.png)

**Comment by brainchild0** [comment]
Thanks for the explanation, and especially, the images. It helped me understand the structure.

I think, for the issue of pruning, it might be helpful to consider the following two issues:

1. During a potential pruning operation, how may the persistent structures be manipulated to clear space (e.g. whole pages) for freeing, without altering documents not selected for pruning?
2. How may new revisions be committed such as to minimize the expected incidence of fragmentation due to pruning, in case  pruning operations may be invoked at a future time?

Considering these questions, the following two even broader questions come to my mind:

1. What loss of storage efficiency, if any, is acceptable (e.g. incomplete use of pages)?
1. What changes to the storage format, if any, are acceptable?

For my part, I have two specific questions, as follows:

1. Are all pages the same fixed size?
2. Are offsets represented as relative to the page, or to the file?

**Comment by JohannesLichtenberger** [comment]
1. The pages are are variable sized and only word aligned. `RevisionRootPages` are the second to last page written and currently aligned to a multiple of 512 bytes. The first two pages (the header/UberPage) replicated once in both files are the only pages, which are overwritten every time a new a new revision is appended.
2. The offsets are absolute and tracked in the `PageReference`s pointing to the leaf page. First thing in a page then is the variable length of the page (currently Snappy compressed/don't know why LZ4 is not working, that's another issue besides a few data dependent simple compressions).

The concept of the sliding snapshot algorithm is described in Marc's thesis: http://kops.uni-konstanz.de/bitstream/handle/123456789/27695/Kramis_276959.pdf

I think that's also why we'd have to write into a temp resource, delete the old resource once no trx is opened and to rename the directory to the old resource name. SirixDB also implements incremental, differential and full versioning of the leaf data pages (with incremental and differential writing a full snapshot of a data page after a predefined number of modifications/revisions). Thus, full versioning for sure is the simplest as it copies the whole page, also in terms of pruning, but we should support all variants, where's almost always the sliding snapshot should be a good middle ground (to write less data and avoid write amplification and unpredictable write peaks due to intermittent full copy-on-writes of a page, if a page is heavily modified).

**Comment by brainchild0** [comment]
It would be nice if pruning were possible without a full rewrite, but it is still helpful for support with the temporary resource, because otherwise an application is forced to accumulate capacity requirements indefinitely. For a data center the problem may not be important, but on embedded targets it is potentially   severe.

If you think of a clever idea, for more integrated support, I would be happy to learn of  it. I am not familiar with many of the concepts you mention, including    the one referenced in the thesis document, and unfortunately, it would not be feasible for me to learn in the foreseeable future.

**Comment by JohannesLichtenberger** [comment]
You could also check if it really requires a lot of storage space. In my tests so far adding or changing some data usually results in minor additional space (as the pages/page fragments are variable sized and not simply copied in 4kb or even 16 kb blocks). I hoped for Intel Optane Byte Addressable Persistent Memory to get successful, but hmm... it would have made fetching 4 cache lines (256 bytes) in the lowest granularity really performant latency wise. Thus, a high end PCIe 4 SSD currently is still best suited (reading small sized page fragments in parallel from differebt offsets in the file to reconstruct a full page in-memory if a page/node gets altered a lot of times).

**Comment by JohannesLichtenberger** [comment]
Regarding the versioning modes think of backup software (incremental, differential and full) but on a per node level instead of file-level.

**Comment by brainchild0** [comment]
> The sliding snapshot algorithm is a novel addition to _not_ store a long chain of incremental deltas of a page with intermittant full snapshots of the page for instance.

In other words, a page view is reconstructed through a hierarchy of reference to fragments spread across multiple physical pages, such that a new page view may be reconstructed on-the-fly through CoW abstraction?

**Comment by JohannesLichtenberger** [comment]
Basically the inner page to data page reference has a list of page fragments, which has at most a size of N (N usually is relatively low). Thus, if nodes in a page have been deleted or updated in N+1 revisions the algorithm copies all unchanged records from the page in current revision - N, too. Thus, always at maximum N page fragments have to be read. If the page has only been modified in one revision the page fragment is the full page for instance. In order to reconstruct the full page the page fragments are read in descending order (regarding the revision number) and only nodes, which are not set are going to be set in each iteration over a new page fragment.

**Comment by JohannesLichtenberger** [comment]
It's described in only two pages in http://kops.uni-konstanz.de/bitstream/handle/123456789/27695/Kramis_276959.pdf  (page 29 and 30). The diagram on page 30 and the description might further help to understand what's going on. The cells are for instance the nodes.

**Comment by brainchild0** [comment]
I will try to review the explanation of the sliding snapshot, when possible.

Meanwhile, I tend to understanding introduction of a     pruning feature by the following existing limitations:

1. The main data structure, of a prefix tree with a time axis, is persisted over a lower-level paged store, which has no current support for defragmentation. 
2. References are given relative to the entire resource, not as page-offset pairs.
3. Pages are arbitrary size.
4. The physical representation of nodes for any revision may be shared in general with the representations for nodes of the immediately preceding revision, as well as any earlier revision without constraint. 

Items (1), (2), and (3) appear to be low-level storage issues.

I believe that item (4), and perhaps to some degree the others, may be resolved by the [earlier suggestion](https://github.com/sirixdb/sirix/issues/562#issuecomment-1334661858). The idea was to constrain dependence of physical representations across revisions, through  considering each revision as a node on a tree populated breadth first, such that stored elements may be shared only with tree ancestors.

**Comment by JohannesLichtenberger** [comment]
I'm not sure how this is supposed to work. Maybe you can elaborate on this further and make a diagram? The delta approach you mentioned is only one of the approaches SirixDB is using and IMHO not the best choice in most cases (differential copy-on-write of the page, which is adding all currently changed nodes as well as all nodes which have been added/changed/deleted since the last full snapshot of the page).

**Comment by brainchild0** [comment]
Right. 

The discussion is complicated by the multiple layers of the storage model. At top is the prefix tree with time axis, and below is a copy-on-write page storage model. Correct?

Both layers would need to integrate the concept, somehow, for it to be meaningful. Yet, as a concept, at least, I think it is valid, that any physical data element may be freed if 1) the revision for which it was originally created is pruned, and 2) no other revision not yet pruned depends on it. 

There is no concrete proposal, at the moment, but I thought that the numbered list might attempt to represent the range of separate questions relating to the broader subject.

**Comment by JohannesLichtenberger** [comment]
The thing is that I don't think we can come up with anything except for rewriting the resource, starting from a given point in time or revision. Otherwise, we'd not be able to for instance reconstruct older revisions, as a page may for instance be altered in almost every revision (depending on the use case for sure).

The revision offsets and timestamps are stored in the second file, which point to the `RevisionRootPage`s. The revision root pages are the main entry point into the storage at a  given revision. The trie itself, which indexes the nodes (and maybe also secondary indexes stored in other tries, a path summary, a dictionary for strings (element- and attribute-names or in JSON object fields).

**Comment by brainchild0** [comment]
Rewriting the resource would be nice, if another solution is not possible.

I was trying to explore other possibilities. I thought it was a good start, but I don't know enough to make a robust argument.

**Comment by JohannesLichtenberger** [comment]
As the system is heavily inspired by ZFS (which copies the whole blocks on write in contrast to SirixDB), we could maybe free space also by looking at the birth time of the pages. I'll have to think about it, but we'd also need an additional virtual table for the offsets in the file... not sure if it's feasible.

https://youtu.be/NXg86uBDSqI

If you watch this you'll get an understanding of the copy-on-write stuff. Additionally, SirixDB references up to N page fragments from the inner page to a leaf data page referencing several versions of the same page in older versions to reduce on-disk space consumption and write amplification due to the copy-on-write overhead. If the page is never modified again after the initial creation only one reference exists for sure and new revisions simply point to this page :-)

**Comment by brainchild0** [comment]
Is the low-level CoW storage system cleanly separable from the data structures?  Is there any possibility for making the low-level store a configurable option, for example, adding one that is easier to defragment?

Does the CoW method add considerable savings in space requirements, beyond the efficiency provided by the prefix tree?

**Comment by JohannesLichtenberger** [comment]
Sure, you can configure for instance full copies of pages. However, a pruning method has to work for all configurations IMHO.

You could also try for yourself how much storage it needs and do some tests?

**Comment by brainchild0** [comment]
> As the system is heavily inspired by ZFS (which copies the whole blocks on write in contrast to SirixDB), we could maybe free space also by looking at the birth time of the pages. I'll have to think about it, but we'd also need an additional virtual table for the offsets in the file... not sure if it's feasible.

What makes the virtual table difficult? Also, for freeing pages, I think it not the birth time that matters, but reference count, right

**Comment by brainchild0** [comment]
> Sure, you can configure for instance full copies of pages. However, a pruning method has to work for all configurations IMHO.

Why? You can add a option called `prunable`, and an option for low-level store (if not already existing). Then, if a conflict is detected at the point of initiating the persistent resource,  fail.

**Comment by JohannesLichtenberger** [comment]
> > As the system is heavily inspired by ZFS (which copies the whole blocks on write in contrast to SirixDB), we could maybe free space also by looking at the birth time of the pages. I'll have to think about it, but we'd also need an additional virtual table for the offsets in the file... not sure if it's feasible.
> 
> What makes the virtual table difficult? Also, for freeing pages, I think it not the birth time that matters, but reference count, right?

It adds indirection (the table lookup, which might or might not be significant). Currently, I'm focusing on better write- and read latency :-)

There is no reference count we could check for a given revision (for instance we don't want to modify pages in-place). Thus, it's not known how many times a page is references from future revisions.

However at this point, I'd also suggest simply doing some tests if storage consumption really is an issue (maybe very infrequently add a new resource and delete the old resource is even an option).

**Comment by brainchild0** [comment]
> However, at this point, I'd also suggest simply doing some tests if storage consumption is an issue (maybe very infrequently, adding a new resource and deleting the old resource is even an option).

Well, the issue is that space requirements expand indefinitely. If the number of revisions is large compared to the size of the document, then such unlimited growth will tend not to be acceptable. I understand it is not a problem for some environments, but consumer devices are different from data centers.

**Comment by brainchild0** [comment]
> It adds indirection (the table lookup, which might or might not be significant). Currently, I'm focusing on better write- and read- latency :-)

Is your concern that address mapping increases latency?


> 
> There is no reference count we could check for a given revision (for instance, we don't want to modify pages in-place). Thus, it's not known how many times a page is referenced from future revisions.

Can you add one?

**Comment by brainchild0** [comment]
> (maybe very infrequently, adding a new resource and deleting the old resource is even an option).

Yes, it is possible in principle.

**Comment by JohannesLichtenberger** [comment]
I don't think it will be a problem, as usually the revision deltas (the data you add/update/delete is rather small). However, I think we could provide a mechanism as I've sketched a few comments ago). The number of times a page is referenced cannot be changed as it would involve an additional in-place modification (which I deliberately don't want to do, because of SSD characteristics and the added read/write of the page).

Compared to a non-temporal system the additional storage space is largely dependent on how many times you'd update nodes. All other variants just add data pages (and maybe one or a few kb for writing the ancestor chain up to the `RevisionRootPage`).

Despite, I'd still suggest to make some tests and let me know how much storage space is used :-)

**Comment by brainchild0** [comment]
> However, I think we could provide a mechanism as I've sketched a few comments ago). 

Are you discussing rebuilding a new resource through an internal process? Yes, I believe it would be quite helpful.

> The number of times a page is referenced cannot be changed as it would involve an additional in-place modification (which I deliberately don't want to do, because of SSD characteristics and the added read/write of the page).

I understand the design problems of keeping a reference count of a page inside of the page itself, but plainly other locations for storing the data are more compatible with the physical attributes of the media, for example, keeping the count data in header, footer, or special intermediary pages, which would  consolidate the various data fields frequently updated for the whole resource.


> 
> Compared to a non-temporal system the additional storage space is largely dependent on how many times you'd update nodes. All other variants just add data pages (and maybe one or a few kb for writing the ancestor chain up to the `RevisionRootPage`).
> 


> I don't think it will be a problem, as usually the revision deltas (the data you add/update/delete is rather small). 


> Despite, I'd still suggest to make some tests and let me know how much storage space is used :-)


Yes, I understand all of it. What emerges are two conceptual observations.

First unbounded growth within a bounded capacity is not sustainable. Second, as  the number of revisions becomes increasingly significant compared to the 
size of  the document (, the relevance of sharing nodes across revisions diminishes, with respect to the problem of managing storage requirements.

**Comment by JohannesLichtenberger** [comment]
We'll tackle this in #589
